cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java

```diff
@@ -479,6 +497,8 @@ public void updateApp(FullHttpRequest request, HttpResponder responder,
     }
 
     try {
+      String latestAppVersion = getLatestAppVersion(appId);
+      appId = validateApplicationVersionId(namespaceId, appName, latestAppVersion);
```

Comment:
Overall it's a bit weird doing this in the handler outside of the transactional context. I feel there are concurrency risks here. "Latest" version mapping should happen in the service as part of the transaction

Reply:
I agree - and that happens here in the deployment https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/DefaultStore.java#L431 - Should we not change this api instead? Because if we include a check for latest version instead - it breaks existing behavior?

Reply:
I think we should not change this api for now - we should probably discuss this with product and check how we can change its behavior - since this api updates app-spec of an existing app (we should probably not create a new version)?

Reply:
It's not about API change. It's just the version retrieval should happen inside the service in transaction, not inside the handler. Generally, handlers are not supposed to have much logic.

Reply:
this is a bigger change, tracking it here https://cdap.atlassian.net/browse/CDAP-19782

Reply:
Agree that this is a race condition. In fact, should we even support update app? With LCM, every version is actually immutable. It was true even before the LCM, with -SNAPSHOT as the special case.